System.Text.Json does not correctly handle nested inner types of a generic outer type.

### Description

I'm seeing an `ArgumentOutOfRange` exception when attempting to use the System.Text.Json source generator in my project. This appears to be when a field or property of a type to be generated is of the form `OuterType<T>.Inner`.

### Reproduction Steps

Attempt to compile the following code:

```cs
using System.Text.Json.Serialization;

public class Outer<T>
{
    public class Inner { }
}

public class C
{
    public Outer<int>.Inner I { get; set; }
}

[JsonSerializable(typeof(C))]
public partial class Context : JsonSerializerContext
{
    public static string Implicit => "World";
}
```

### Expected behavior

The code compiles.

### Actual behavior

```cs
CSC : warning CS8785: Generator 'JsonSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentOutOfRangeException' with message 'Length cannot be less than zero. (Parameter 'length')' [C:\Users\333fr\Documents\git\omnisharp-roslyn\src\OmniSharp.Abstractions\OmniSharp.Abstractions.csproj]
```

Followed by errors from lack of partial type implementation.

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

.NET 6.0.100.

### Other information

_No response_

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs
@@ -126,6 +126,18 @@ namespace System.Text.Json.SourceGeneration.Tests
     {
         [JsonSerializable(typeof(JsonMessage))]
         public partial class NestedInGenericContainerContext : JsonSerializerContext { }
+
+        [JsonSerializable(typeof(JsonMessage))]
+        public partial class NestedGenericInGenericContainerContext<T1> : JsonSerializerContext { }
+
+        public partial class NestedGenericContainer<T1>
+        {
+            [JsonSerializable(typeof(JsonMessage))]
+            public partial class NestedInNestedGenericContainerContext : JsonSerializerContext { }
+
+            [JsonSerializable(typeof(JsonMessage))]
+            public partial class NestedGenericInNestedGenericContainerContext<T2> : JsonSerializerContext { }
+        }
     }
 
     [JsonSerializable(typeof(MyContainingClass.MyNestedClass.MyNestedNestedClass))]
@@ -136,5 +148,6 @@ namespace System.Text.Json.SourceGeneration.Tests
     [JsonSerializable(typeof(MyContainingGenericClass<int>.MyNestedClass.MyNestedNestedGenericClass<int>))]
     [JsonSerializable(typeof(MyContainingGenericClass<int>.MyNestedGenericClass<int>.MyNestedGenericNestedClass))]
     [JsonSerializable(typeof(MyContainingGenericClass<int>.MyNestedGenericClass<int>.MyNestedGenericNestedGenericClass<int>))]
+    [JsonSerializable(typeof(MyContainingGenericClass<MyContainingGenericClass<int>.MyNestedGenericClass<int>.MyNestedGenericNestedGenericClass<int>>.MyNestedGenericClass<int>.MyNestedGenericNestedGenericClass<int>))]
     internal partial class NestedGenericTypesContext : JsonSerializerContext { }
 }

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs
@@ -24,27 +24,31 @@ namespace System.Text.Json.SourceGeneration.Tests
         [Fact]
         public static void VariousGenericsAreSupported()
         {
-            Assert.NotNull(GenericContext<int>.Default);
-            AssertSerialization(GenericContext<int>.Default.JsonMessage);
-
-            Assert.NotNull(ContextGenericContainer<int>.NestedInGenericContainerContext.Default);
-            AssertSerialization(ContextGenericContainer<int>.NestedInGenericContainerContext.Default.JsonMessage);
-
-            Assert.NotNull(ContextGenericContainer<int>.NestedGenericInGenericContainerContext<int>.Default);
-            AssertSerialization(ContextGenericContainer<int>.NestedGenericInGenericContainerContext<int>.Default.JsonMessage);
-
-            Assert.NotNull(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedInNestedGenericContainerContext.Default);
-            AssertSerialization(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedInNestedGenericContainerContext.Default.JsonMessage);
-
-            Assert.NotNull(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedGenericInNestedGenericContainerContext<int>.Default);
-            AssertSerialization(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedGenericInNestedGenericContainerContext<int>.Default.JsonMessage);
+            AssertGenericContext(GenericContext<int>.Default);
+            AssertGenericContext(ContextGenericContainer<int>.NestedInGenericContainerContext.Default);
+            AssertGenericContext(ContextGenericContainer<int>.NestedGenericInGenericContainerContext<int>.Default);
+            AssertGenericContext(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedInNestedGenericContainerContext.Default);
+            AssertGenericContext(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedGenericInNestedGenericContainerContext<int>.Default);
 
             Assert.NotNull(NestedGenericTypesContext.Default);
-
-            static void AssertSerialization(JsonTypeInfo<JsonMessage> jsonTypeInfo)
+            var original = new MyContainingGenericClass<int>.MyNestedGenericClass<int>.MyNestedGenericNestedGenericClass<int>()
             {
-                string json = JsonSerializer.Serialize(new JsonMessage { Message = "Hi" }, jsonTypeInfo);
-                JsonMessage deserialized = JsonSerializer.Deserialize<JsonMessage>(json, jsonTypeInfo);
+                DataT = 1,
+                DataT1 = 10,
+                DataT2 = 100
+            };
+            Type type = typeof(MyContainingGenericClass<int>.MyNestedGenericClass<int>.MyNestedGenericNestedGenericClass<int>);
+            string json = JsonSerializer.Serialize(original, type, NestedGenericTypesContext.Default);
+            var deserialized = (MyContainingGenericClass<int>.MyNestedGenericClass<int>.MyNestedGenericNestedGenericClass<int>)JsonSerializer.Deserialize(json, type, NestedGenericTypesContext.Default);
+            Assert.Equal(1, deserialized.DataT);
+            Assert.Equal(10, deserialized.DataT1);
+            Assert.Equal(100, deserialized.DataT2);
+
+            static void AssertGenericContext(JsonSerializerContext context)
+            {
+                Assert.NotNull(context);
+                string json = JsonSerializer.Serialize(new JsonMessage { Message = "Hi" }, typeof(JsonMessage), context);
+                JsonMessage deserialized = (JsonMessage)JsonSerializer.Deserialize(json, typeof(JsonMessage), context);
                 Assert.Equal("Hi", deserialized.Message);
             }
         }

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/JsonSerializerContextTests.cs
@@ -24,9 +24,29 @@ namespace System.Text.Json.SourceGeneration.Tests
         [Fact]
         public static void VariousGenericsAreSupported()
         {
-            Assert.NotNull(GenericContext<object>.Default);
-            Assert.NotNull(ContextGenericContainer<object>.NestedInGenericContainerContext.Default);
+            Assert.NotNull(GenericContext<int>.Default);
+            AssertSerialization(GenericContext<int>.Default.JsonMessage);
+
+            Assert.NotNull(ContextGenericContainer<int>.NestedInGenericContainerContext.Default);
+            AssertSerialization(ContextGenericContainer<int>.NestedInGenericContainerContext.Default.JsonMessage);
+
+            Assert.NotNull(ContextGenericContainer<int>.NestedGenericInGenericContainerContext<int>.Default);
+            AssertSerialization(ContextGenericContainer<int>.NestedGenericInGenericContainerContext<int>.Default.JsonMessage);
+
+            Assert.NotNull(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedInNestedGenericContainerContext.Default);
+            AssertSerialization(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedInNestedGenericContainerContext.Default.JsonMessage);
+
+            Assert.NotNull(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedGenericInNestedGenericContainerContext<int>.Default);
+            AssertSerialization(ContextGenericContainer<int>.NestedGenericContainer<int>.NestedGenericInNestedGenericContainerContext<int>.Default.JsonMessage);
+
             Assert.NotNull(NestedGenericTypesContext.Default);
+
+            static void AssertSerialization(JsonTypeInfo<JsonMessage> jsonTypeInfo)
+            {
+                string json = JsonSerializer.Serialize(new JsonMessage { Message = "Hi" }, jsonTypeInfo);
+                JsonMessage deserialized = JsonSerializer.Deserialize<JsonMessage>(json, jsonTypeInfo);
+                Assert.Equal("Hi", deserialized.Message);
+            }
         }
 
         [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]

--- a/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs
+++ b/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs
@@ -218,7 +218,12 @@ namespace System.Text.Json.SourceGeneration.Tests
         public class MyNestedGenericClass<T1>
         {
             public class MyNestedGenericNestedClass { }
-            public class MyNestedGenericNestedGenericClass<T2> { }
+            public class MyNestedGenericNestedGenericClass<T2>
+            {
+                public T DataT { get; set; }
+                public T1 DataT1 { get; set; }
+                public T2 DataT2 { get; set; }
+            }
         }
     }
 }

